### PR TITLE
[vLLM IR] Port activations (gelu)  to IR op

### PR DESCRIPTION
**Pull request by Alex-ai-future**
## Purpose

Migrate GELU activations from CustomOp to PluggableLayer

#### Port Changes

Migrated 5 GELU-series activation functions from `CustomOp` to `PluggableLayer`:

| Class | Before | After |
|---|---|---|
| `GELU` | `CustomOp` | `PluggableLayer` → `ir.ops.gelu` |
| `GeluAndMul` | `CustomOp` | `PluggableLayer` → `ir.ops.gelu_and_mul` |
| `NewGELU` | `CustomOp` | `PluggableLayer` → `ir.ops.gelu_new` |
| `FastGELU` | `CustomOp` | `PluggableLayer` → `ir.ops.gelu_fast` |
| `QuickGELU` | `CustomOp` | `PluggableLayer` → `ir.ops.quick_gelu` |

Each layer is simplified from ~40 lines (with `forward_cpu`, `forward_cuda`, `forward_xpu` platform branches) to ~5 lines (single `forward()` calling `ir.ops.*`).

---

#### Test Changes

**Test logic has undergone significant restructuring.**

**Before — the problem:**
The old tests mixed two concerns:
1. **Kernel output correctness** — verifying the CUDA kernel produces numerically correct results
2. **Routing verification** — verifying the PluggableLayer dispatches to the correct `ir.ops` function

This resulted in tests with many parameter combinations (`num_tokens`, `d`, `dtype`, `seed`, `device`) that were testing both concerns simultaneously, creating tight coupling to platform-specific details.

**After — the new approach:**

**2.1 Simplified tests: routing only, with reusable infrastructure**

The tests in `test_activation.py` now focus on **one concern only**: verifying that PluggableLayer instances route to the correct `ir.ops` function.

```python
ACTIVATION_LAYER_CONFIGS = [
    ("GELU", GELU, ir.ops.gelu, {}),
    ("GeluAndMul_none", GeluAndMul, ir.ops.gelu_and_mul, {"approximate": "none"}),
    ...
]

@pytest.mark.parametrize("name, layer_cls, ir_op, kwargs", ACTIVATION_LAYER_CONFIGS)
def test_activation_ir_op_routing(...):
    """Verify PluggableLayer calls the corresponding ir.ops function."""
```

- Removed all parameter combinations (`num_tokens`, `d`, `dtype`, `seed`, `device`)
- Uses a fixed input tensor since routing correctness is independent of input shape/dtype
- Added `assert_pluggable_layer_calls_ir_op()` in `tests/kernels/utils.py` as reusable infrastructure for all PluggableLayer routing tests

**2.2 Reduced parameter dependencies — to be covered by `ir.ops` tests**

Parameter variations (`dtype`, `shape`, etc.) and kernel output correctness verification will be tested in dedicated `ir.ops` unit tests (under `tests/kernels/ir/`). This separates concerns:
- `test_activation.py` → routing logic
- `tests/kernels/ir/` → kernel implementation correctness

Also removed GELU tests from `test_cpu_activation.py` since GELU kernels are CUDA-only (`CUDA_ALIKE`).

---

#### Remaining Work

- [ ] **Add `ir.ops` unit tests**: Current tests verify routing but not kernel output correctness. Need to add tests under `tests/kernels/ir/` that:
  - Verify each `ir.ops` kernel output matches PyTorch native implementation
  - Cover multiple dtypes (`float16`, `bfloat16`, `float32`)
  - Cover multiple shapes/tensor sizes
---
## Related

- https://github.com/vllm-project/vllm/issues/38733
- https://github.com/vllm-project/vllm/issues/32676
- https://github.com/vllm-project/vllm/pull/33825
---
- Corrections and feedback are welcome.

## Test Plan
```
 pytest tests/kernels/core/test_cpu_activation.py tests/kernels/core/test_activation.py tests/model_executor/test_enabled_custom_ops.py -v
```

## Test Result

```
=================================== 117 passed, 7 skipped, 16 warnings in 33.93s ===================================
```
---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

**Comment by Alex-ai-future** [comment]
cc @ProExpertProg

File: vllm/ir/ops/activation.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
+import math
+
+import torch
+from torch import Tensor
+
+from ..op import register_op
+
+c_gelu_new = math.sqrt(2.0 / math.pi)
```
Comment (ProExpertProg) [inline vllm/ir/ops/activation.py:10]: Call this `SQRT_2_OVER_PI`?

File: tests/kernels/ir/test_activation.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        gelu_op = getattr(ir.ops, op_name)
+        actual = {
+            provider: impl.supported
+            for provider, impl in gelu_op.impls.items()
+        }
+        assert actual == expected, f"{op_name} has incorrect registration"
+
+
+@pytest.mark.parametrize("dtype", [torch.float16, torch.bfloat16, torch.float32])
+@pytest.mark.parametrize("shape", [(1, 8), (4, 16), (17, 64)])
```
Comment (ProExpertProg) [inline tests/kernels/ir/test_activation.py]: Please use the new helpers (see layernorm test)

File: tests/kernels/ir/test_activation.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+class TestGeluOps:
+    """Tests for GELU IR ops."""
+
+    @pytest.mark.parametrize(
+        "gelu_op_name", ["gelu_new", "gelu_fast", "quick_gelu"]
```
Comment (ProExpertProg) [inline tests/kernels/ir/test_activation.py]: Can this be a class param?

File: vllm/ir/ops/activation.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+        current_platform.is_cpu()
+        and current_platform.get_cpu_architecture() == CpuArchEnum.ARM
+        and x.dtype == torch.bfloat16
+        and x.is_contiguous()
+        and hasattr(torch.ops._C, "activation_lut_bf16")
+    )
+    if is_arm_bf16:
+        out = torch.empty_like(x)
+        torch.ops._C.activation_lut_bf16(out, x, "gelu")
+        return out
```
Comment (ProExpertProg) [inline vllm/ir/ops/activation.py:37]: Register this as an implementation for gelu op and specify it in the priority list

File: vllm/ir/ops/activation.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    d = x.shape[-1] // 2
+    return F.gelu(x[..., :d], approximate=approximate) * x[..., d:]
+
+
+@gelu_and_mul.register_input_generator
+def _gelu_and_mul_input_generator(
+    num_tokens: int, hidden_size: int, dtype: torch.dtype
+) -> tuple:
+    # hidden_size must be even for gelu_and_mul (it's 2*d)
+    x = torch.randn(num_tokens, hidden_size, dtype=dtype)
```
Comment (ProExpertProg) [inline vllm/ir/ops/activation.py:72]: Return 2*hidden_size then!

File: vllm/kernels/vllm_c.py [hunk 6]
```diff
@@ -31,3 +31,63 @@ def rms_norm(
...
+
+
+#===================
+# GELU Activations
+#===================
+
+
+@ir.ops.gelu_new.register_impl("vllm_c", supported=CUDA_ALIKE)
+def gelu_new(x: Tensor) -> Tensor:
+    """
```
Comment (ProExpertProg) [inline vllm/kernels/vllm_c.py]: These are a bit verbose, just leave the formula

File: vllm/ir/ops/activation.py [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+@gelu.register_input_generator
+def _gelu_input_generator(
+    num_tokens: int, hidden_size: int, dtype: torch.dtype
+) -> tuple:
+    x = torch.randn(num_tokens, hidden_size, dtype=dtype)
+    return (x,)
+
+
+@register_op
+def gelu_and_mul(x: Tensor, approximate: str = "none") -> Tensor:
```
Comment (ProExpertProg) [inline vllm/ir/ops/activation.py:52]: Can we type-constrain the `approximate` param to a `Literal[...]` with the allowed choices?

File: vllm/config/kernel.py [hunk 8]
```diff
@@ -31,6 +31,21 @@ class IrOpPriorityConfig:
     rms_norm: list[str] = Field(default_factory=list)
     """Priority list for vllm.ir.ops.rms_norm"""
 
+    gelu: list[str] = Field(default_factory=list)
```
Comment (ProExpertProg) [inline vllm/config/kernel.py:37]: I'm realizing these docstrings are redundant, let's just collapse these to one op per line and no spaces (and group activations together)

File: tests/model_executor/test_enabled_custom_ops.py [hunk 9]
```diff
@@ -44,39 +43,39 @@ class Relu3(ReLUSquaredActivation):
...
         #
         # Only ReLU3
-        ("none,+relu3", 3, "inductor", [0, 0, 0, 1], False),
+        ("none,+relu3", 3, "inductor", [0, 0, 1], False),
         # All but RMSNorm
-        ("all,-rms_norm", 3, "inductor", [0, 1, 1, 1], True),
+        ("all,-rms_norm", 3, "inductor", [0, 1, 1], True),
     ],
 )
 def test_enabled_ops(
```
Comment (ProExpertProg) [inline tests/model_executor/test_enabled_custom_ops.py:91]: Switch to using `CustomOp` subclasses defined in the test, as we remove more and more actual `CustomOp`s (eventually leading to full deprecation)

File: tests/kernels/utils.py [hunk 10]
```diff
@@ -1085,3 +1085,36 @@ def group_broadcast(t, shape):
         output = output + bias
 
     return output
+
+
+def assert_pluggable_layer_calls_ir_op(
```
Comment (ProExpertProg) [inline tests/kernels/utils.py]: We don't need this because we're gonna get rid of the pluggable layers anyway

File: tests/kernels/core/test_cpu_activation.py [hunk 11]
```diff
@@ -1,25 +1,25 @@
 # SPDX-License-Identifier: Apache-2.0
 # SPDX-FileCopyrightText: Copyright contributors to the vLLM project
 
+# NOTE: This file contains CPU-only activation kernel tests.
```
Comment (ProExpertProg) [inline tests/kernels/core/test_cpu_activation.py]: Note this will be removed once we remove `SiluAndMul`

File: tests/kernels/utils.py (hunk 10 above)
Comment (Alex-ai-future) [reply tests/kernels/utils.py]: Are you suggesting that, in the future, when we are developing models, we will directly use `ir.op`—effectively turning it into the language used to describe the models?

**Comment by whkuo269** [comment]
Hi @Alex-ai-future , @ProExpertProg , 

I took a look at this draft and the remaining work around GELU IR op correctness tests.

It looks like the current tests mainly cover `PluggableLayer` -> `ir.ops` routing, while dedicated tests under `tests/kernels/ir/` for GELU-family IR op correctness are still missing, as listed in the PR description.

I can help with a small focused patch that adds `tests/kernels/ir/test_activation.py` following the existing IR op test patterns from `test_mul_and_silu.py` / `test_relu2.py`:

- native semantics checks against PyTorch/reference formulas
- provider implementation parity checks for supported providers, e.g. `vllm_c`
- dispatch priority checks comparing dispatched output with direct provider output
- dtype/shape coverage for `float16`, `bfloat16`, `float32`

I’m happy to prepare this either as a small patch for this draft or as a follow-up PR, whichever is easier for you.

**Comment by Alex-ai-future** [comment]
> Hi @Alex-ai-future , @ProExpertProg , 
> 
> 
> 
> I took a look at this draft and the remaining work around GELU IR op correctness tests.
> 
> 
> 
> It looks like the current tests mainly cover `PluggableLayer` -> `ir.ops` routing, while dedicated tests under `tests/kernels/ir/` for GELU-family IR op correctness are still missing, as listed in the PR description.
> 
> 
> 
> I can help with a small focused patch that adds `tests/kernels/ir/test_activation.py` following the existing IR op test patterns from `test_mul_and_silu.py` / `test_relu2.py`:
> 
> 
> 
> - native semantics checks against PyTorch/reference formulas
> 
> - provider implementation parity checks for supported providers, e.g. `vllm_c`
> 
> - dispatch priority checks comparing dispatched output with direct provider output
> 
> - dtype/shape coverage for `float16`, `bfloat16`, `float32`
> 
> 
> 
> I’m happy to prepare this either as a small patch for this draft or as a follow-up PR, whichever is easier for you.
> 
> 
> 
> 

At present, it seems that this work needs to be re-planned.

**Comment by whkuo269** [comment]
Hi @Alex-ai-future,

Got it, thanks for the update.

I’ll hold off on preparing the GELU IR correctness test patch until the direction is clearer.

Please let me know if there is a specific direction or task where I can help.
